runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/logging/JulHandlerPrintStreamAdapterFactory.java

```diff
@@ -86,7 +86,7 @@ public void write(int i) throws IOException {
       this.logger = Logger.getLogger(loggerName);
       this.buffer = new StringBuilder();
       this.decoder =
-          Charset.defaultCharset()
+          StandardCharsets.UTF_8
```

Comment:
@scwhittle I was wondering if you could comment on this since you added this line in https://github.com/apache/beam/pull/11096. Do you know if this change is safe? Is there a reason this needs to use `defaultCharset`?

As it is, `testLogRawBytes` below can fail if the system encoding isn't UTF-8

Reply:
This class is going to be used to replace the default System.out/System.err which appears to be specified to use the default charset.
So it seems better to keep it consistent IMO to avoid surprises when someone expects that charset writing to System.out, not UTF-8.  If the test is incorrect, it should be fixed (by examining or setting the defaultCharset)

Reply:
OK thanks, I thought that may be the case

How about this - we can make this class (the private constructor and the create() method below) parameterized by Charset. The places where its used as the default System.out/System.err in the Dataflow worker can pass in Charset.defaultCharset(), but the test can pass in Charsets.UTF_8

Reply:
@scwhittle  does that work for you? I can make the change if so!

Reply:
Let's go ahead and do that

Reply:
Sorry lost track of that comment, I believe it's only uses are for System.out/System.err and the test. But allowing the test to specify a specific charset so it passes reliably sounds like a good solution.

Reply:
OK, I will make the changes and re-commit. Thanks!

Reply:
This hasn't been done yet has it?

Reply:
> How about this - we can make this class (the private constructor and the create() method below) parameterized by Charset. 

The variables that are passed in do not use any of the `Charset` methods though right?



>The places where its used as the default System.out/System.err in the Dataflow worker can pass in Charset.defaultCharset(), but the test can pass in Charsets.UTF_8
The test file has `StandardCharsets.UTF_8` replacing `Charset.defaultCharset()`. 

Let me know if I am missing something @TheNeuralBit !

Reply:
What I was suggesting is that we should add a new argument to the private constructor and the static create() method: `Charset charset`. Then in the constructor we'll use `charset` instead of `Charset.defaultCharset()` to create `encoder`.

Currently the test is using `StandardCharsets.UTF_8`, but only to create the expected outputs. It also needs to create a `JulHandlerPrintStreamAdapterFactory` that uses `UTF_8` explicitly, so we need a way to pass that through. I think as it is this test could still fail if the default charset isn't UTF-8

Reply:
Oooh, OK! And because the constructor is private, it can only be accessed from `create()` method inside the class, meaning this is the only file I have to change 🤯

Ill give this a shot!

Reply:
Well you'll also want to add an argument to `create()`, and update the places that call it, for example in the test: https://github.com/apache/beam/blob/df74d74ef295aa972b0d1ddef2429f427b988a25/runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/logging/JulHandlerPrintStreamAdapterFactoryTest.java#L168

Would become `return JulHandlerPrintStreamAdapterFactory.create(handler, LOGGER_NAME, Level.INFO, StandardCharsets.UTF_8);`

Reply:
Will do, thanks!

Reply:
@TheNeuralBit  I had to create a new PR for this as this got merged: 

New PR: https://github.com/apache/beam/pull/13701